plutus-metatheory/src/VerifiedCompilation/UCaseOfCase.lagda.md

```diff
@@ -0,0 +1,100 @@
+```
+module VerifiedCompilation.UCaseOfCase where
+
+```
+
+## Imports
+
+```
+open import Data.Nat using (ℕ;zero;suc)
+open import Data.Empty using (⊥)
+open import Scoped using (ScopedTm; Weirdℕ)
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+open import Untyped.CEK using (BApp; fullyAppliedBuiltin; BUILTIN; stepper; State; Stack)
+open import Evaluator.Base using (maxsteps)
+open import Builtin using (Builtin; ifThenElse)
+open import Data.List using (List; zip; [_])
+open import Utils as U using (Maybe;nothing;just;Either)
+open import RawU using (TmCon)
+open import Data.List.Relation.Unary.All using (All; []; _∷_)
+open import Data.Product using (_×_; proj₁; proj₂)
+open import Relation.Nullary using (Dec; yes; no)
+```
+
+## Translation Relation
+
+This compiler stage only applies to the very specific case where an `IfThenElse` builtin exists in a `case` expression.
+It moves the `IfThenElse` outside and creates two `case` expressions with each of the possible lists of cases. 
+
+This translation relation has a constructor for the specific case, and then inductive constructors for everything else
+to traverse the ASTs.
+
+```
+data _⊢̂_⊳̂_ (X : Set) : (X ⊢) → (X ⊢) → Set where
+--   refl : ∀ {X : Set} {x : X ⊢} → X ⊢̂ x ⊳̂ x
+   caseofcase : ∀ {b : X ⊢} {tn tn' fn fn' : ℕ} {alts alts' tt ft tt' ft' : List (X ⊢)} 
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip alts alts') -- recursive translation for the other case patterns 
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip tt tt') -- recursive translation for the true branch
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip ft ft') -- recursive translation for the false branch
+                ----------------------
+                → X ⊢̂
+                       (case ((((force (builtin ifThenElse)) · b) · (constr tn tt)) · (constr fn ft)) alts)
+                       ⊳̂ ((((force (builtin ifThenElse)) · b) · (case (constr tn' tt') alts')) · (case (constr tn' tt') alts'))
```

Comment:
Again, this will evaluate both the branches of `ifThenElse`, i.e. you're going to evaluate two branches of `alts'` in total in general (in a particular case they could both end up being the same branch), while the original term doesn't have behavior (it does evaluate both branches of `ifThenElse`, but there are no `alts` inside of them, just `constr`s).

I.e. I think this translation is not semantics preserving.

Reply:
This Agda is just suppose to represent what the Haskell does - if you think the Haskell is wrong then we should fix that first :)

Reply:
Yep, you're totally right, I completely missed the fact that it's the Haskell code that I find deeply suspicious.

Reply:
This is what Formal Methods are supposed to do - force us to look at the details of stuff we haven't previously :) Let me know what the "right" answer is and I'll make the spec match _that_.

Reply:
> This is what Formal Methods are supposed to do - force us to look at the details of stuff we haven't previously :) 

We did look at it at review time! At the general rule and at examples. I have no idea how we missed it. But let's not get ahead of ourselves, maybe I'm talking gibberish and it's all fine. We'll get to know soon.

> Let me know what the "right" answer is and I'll make the spec match that.

Will do.